src/backings/tree.ts

```diff
@@ -0,0 +1,376 @@
+import {Proof, Tree} from "@chainsafe/persistent-merkle-tree";
+import {CompositeValue, ArrayLike, List} from "../interface";
+import {
+  BasicArrayType,
+  BasicListType,
+  CompositeArrayType,
+  CompositeListType,
+  CompositeType,
+  ContainerType,
+  isBasicType,
+  isCompositeType,
+  isContainerType,
+  isListType,
+  isVectorType,
+} from "../types";
+import {isTree} from "../util/tree";
+import {byteArrayEquals} from "../util/byteArray";
+import {Path} from "./backedValue";
+
+export function isTreeBacked<T extends CompositeValue>(value: unknown): value is ITreeBacked<T> {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  return value && (value as any).type && (value as any).tree && isTree((value as any).tree);
+}
+
+export type KeyOf = string | number;
+export type ValueOf<T extends CompositeValue, V extends keyof T = keyof T> = ITreeBacked<T[V]> | T[V];
+
+export type TreeBacked<T extends CompositeValue> = ITreeBacked<T> & T;
+
+/**
+ * The ITreeBacked interface gives convenient access to a `CompositeType` + `Tree` as a wrapper object
+ *
+ * This is an alternative way of calling `tree_*` methods of the type
+ */
+export interface ITreeBacked<T extends CompositeValue> {
+  type: CompositeType<T>;
+  /**
+   * The merkle tree backing
+   */
+  tree: Tree;
+
+  /**
+   * Equality
+   *
+   * If both values are tree-backed, use equality by merkle root, else structural equality
+   */
+  equals(other: T): boolean;
+  /**
+   * Clone / Copy
+   */
+  clone(): TreeBacked<T>;
+
+  /**
+   * Serialized byte length
+   */
+  size(): number;
+  /**
+   * Low-level serialization
+   *
+   * Serializes to a pre-allocated Uint8Array
+   */
+  toBytes(output: Uint8Array, offset: number): number;
+  /**
+   * Serialization
+   */
+  serialize(): Uint8Array;
+
+  /**
+   * Merkleization
+   */
+  hashTreeRoot(): Uint8Array;
+
+  createProof(paths: Path[]): Proof;
```

Comment:
It would be supper awesome if we could make path typesafe. But it would require some recursive keyof and wouldn't work with arrays right?

Reply:
Yes it does! This compiles

```ts
type State = {
  a: {
    b: {
      c: {
        d: string;
      };
    }[];
  };
};

gett({} as State, "a", "b", 1, "c");

function gett<
  T,
  K1 extends keyof T,
  K2 extends keyof T[K1],
  K3 extends keyof T[K1][K2],
  K4 extends keyof T[K1][K2][K3]
>(val: T, k1: K1, k2: K2, k3: K3, k4: K4) {}
```